### ISSUE TITLE
Phoenix generated code should abide by Credo checks as much as possible

### ISSUE THREAD

**Issue by bmitc** [phoenixframework/phoenix#4881]
### Environment

* Elixir version (elixir -v): n/a
* Phoenix version (mix deps): latest
* Operating system: n/a

### Expected behavior
Phoenix generates a substantial amount of code. However, a decent amount of this code does not pass several Credo checks, which causes issues in codebases that strictly abide by Credo.

### Actual behavior
Phoenix generated code should abide by Credo checks as much as possible, regardless of if the checks are enabled or not so as to be independent of various Credo configurations.

### Suggestion
Phoenix generated code should support the following checks:
1. [Credo.Check.Readability.AliasAs](https://hexdocs.pm/credo/Credo.Check.Readability.AliasAs.html)
    * Thus, code such as:
       ```
       def controller() do
         quote do
           use Phoenix.Controller, namespace: IslandsInterfaceWeb

           import Plug.Conn
           import IslandsInterfaceWeb.Gettext
           alias IslandsInterfaceWeb.Router.Helpers, as: Routes
         end
       end
       ```
       should only use `alias IslandsInterfaceWeb.Router.Helpers`. This would also help maintain compliance with the default configuration of [Credo.Check.Design.AliasUsage](https://hexdocs.pm/credo/Credo.Check.Design.AliasUsage.html)

2. [Credo.Check.Readability.ModuleDoc](https://hexdocs.pm/credo/Credo.Check.Readability.ModuleDoc.html)
    * I think at a minimum, Phoenix generated modules could have documentation such as `@moduledoc false` or `@moduledoc "Phoenix generated module"` to maintain compliance. If possible, actual documentation could be generated. This has no effect other than to pass the check if it is enabled.

3. [Credo.Check.Readability.ImplTrue](https://hexdocs.pm/credo/Credo.Check.Readability.ImplTrue.html)
    * Thus, all behaviour callback implementations should use `@impl <behaviour name>` instead of `@impl true`. This has no effect other than to pass the check if it is enabled.

4. [Credo.Check.Readability.Specs](https://hexdocs.pm/credo/Credo.Check.Readability.Specs.html)
    * Thus, all non-callback implementations, including both public and private functions should have typespecs. At a minimum, the typespecs can be done with `any()`, like `  @spec error_tag(any(), any()) :: any()` or `  @spec index(Plug.Conn.t(), any()) :: any()`. This has no effect other than to pass the check if it is enabled and should not affect Dialyzer.

5. [Optional] This isn't a default Credo check (it's a todo of mine to add a Credo check for this), but all generated functions should ideally have documentation.

6. [Optional] [Credo.Check.Readability.ParenthesesOnZeroArityDefs](https://hexdocs.pm/credo/Credo.Check.Readability.ParenthesesOnZeroArityDefs.html)
    * This is a tough one. I personally feel this should be enforced by the compiler in that only `def test() do` should be allowed and not `def test do`. However, given the situation where either are allowed in Elixir and that there is no way to pick one that is always compliant with the check, there is no great solution. The only one that I can think of is to have an option when generating Phoenix projects that configures the generation to use one or the other. For example, I enforce this check in my codebases, so I would like to tell Phoenix to use `def test() do` everywhere when generating code.
    * The workaround is that one goes through the Phoenix generated code and adds the parentheses. This is not too bad if 1-4 above are fixed.

There was an existing issue that was closed but not resolved: https://github.com/phoenixframework/phoenix/issues/1498. However, that was addressing usability of the generated functions rather than Credo compliance, although there is some overlap. It is possible that this list is not exhaustive. However, I am using Elixir 1.13.4 and Phoenix 1.6.10, and the above checks were the ones that were failing. I don't have an exact count but there were over 30 Credo failures with these checks and strict mode enabled.

I would be willing to work on this if there is agreement. I think 1-4 above are straightforward, since they only provide compliance with the Credo checks if the checks are enabled and have no affect otherwise.

**Comment by josevalim** [comment]
A PR for 3 is definitely welcome.

1 is no longer relevant as verified routes will replace helpers. But note that we think `Routes` is a better name than Helpers, but we can't rename Helpers because it would be a breaking change. So Credo is not giving the flexibility that Phoenix as a framework needs.

I don't think we should do 4. We often aim for generators for learning purposes and I don't think typespecs will help the learning experience.

About 2, it is tricky. I think generating "stubs" for docs is not helpful, because that would be complying to the rules for compliance sake without the benefits of having proper docs. And I am not sure marking everything as `@moduledoc false` is what we want to teach either. In this case, you should either add the docs after generating according to your team rules or remove the check.

**Comment by chrismccord** [comment]
Thanks for the feedback but we will not be moving in this direction.

**Comment by bmitc** [comment]
@josevalim Thanks so much for the feedback! I wouldn't mind working on (3).

For (1), I understand not being able to rename the module and also that `Helpers` is not descriptive. Although, to me this means the module is not named well, but Credo does give flexibility via [`Credo.Check.Design.AliasUsage`](https://hexdocs.pm/credo/Credo.Check.Design.AliasUsage.html) such that `Routers.Helpers` could be used rather than using `alias ..., as: ...`. However, if that module is now going to be obsolete with verified routes (I'm still learning Phoenix, so not sure what that is at the moment, which isn't important), then that sounds good to me. 👍

For (2), is there module documentation that could be useful? I agree that `@moduledoc false` or `@moduledoc """Generated by Phoenix"""` is not too helpful, but it seems to me that generated documentation could be useful, like `@moduledoc """This module is for ____ and can be used or modified to do ____. Place callbacks for ____ module here."""` Just something like that but properly expanded and written with the right help and context. That could both serve as documentation or helper documentation as well as get out of the way of Credo since no one (hopefully!) is going to be enforcing to have no module documentation. Some of the generated Phoenix modules do already have some helper documentation like this in plain comments. I think the existing generated documentation for the `<app_name>` and `<app_name>Web` modules are great examples of what `@moduledoc` could provide for all Phoenix generated modules. For generated test files though, I do think `@moduledoc false` is probably an okay compromise for test modules.

For (4), is that really the case that typespecs wouldn't be useful for learning purposes? 🙂 For example:

```elixir
@spec index(Plug.Conn.t(), map()) :: Plug.Conn.t()
def index(conn, _params) do
  render(conn, "index.html")
end
```
vs
```elixir
def index(conn, _params) do
  render(conn, "index.html")
end
```
I think that's what https://github.com/phoenixframework/phoenix/issues/1498 was trying to address. [I should note, Jose, that you mentioned this there](https://github.com/phoenixframework/phoenix/issues/1498#issuecomment-174433274):
> I agree we should generate `@spec`.

That is unless I misunderstood the context. 😝 (For sure not intending to take your quote out of context, so let me know if I am misunderstanding.) I think it's also maybe addressing the code by macros, which I also agree with. At present, for example, ElixirLS doesn't seem to pick up any typespec for `render/2`, even though [there is one](https://hexdocs.pm/phoenix/Phoenix.Controller.html#render/2). But that may be something else. Not sure.

Of course, this is all revolving around code-styles, but I think *generated* code, as opposed to *library code*, should be as style agnostic (and thus default Credo) as possible. Or another way to put it is to not make waves with Credo failures. The Phoenix codebase may have its own style, but I don't think it should force that style onto users since it generates code. If `@impl`, `@moduledoc`, and `@spec` where properly put in Phoenix generated code, it would greatly reduce the amount of work for people who abide by Credo. For those who don't, it would have no effect. The only potential there is for it to be against their personal style (but it wouldn't fail "official" Credo checks), so they could remove typespecs and the like if they really want to.

I think being as Credo compliant as possible while also providing some benefit and learning to users and then letting devs *subtract* (rather than needing to *add*) from the generated code might be a better process.

---------------------

@chrismccord Does closing this mean a discussion and/or a PR for any of this is not welcome?

**Comment by josevalim** [comment]
My comment about specs was for code generated from macros to help with dialyzer. There is no visual cost.

In any case, following all of Credo rules is a style choice (not syntax style, but a style anyway). All of those things do have a cost, specs can be noisy, template docs that may not be filled in, etc. which is why we are not going forward with this. We generally say that in doubt we prefer to generate less than more, otherwise it is easy to add a lot “just in case”.

Thank you!

**Comment by bmitc** [comment]
Apologies for misunderstanding there. I scratched that bit out so as to not be misleading. I re-read that other PR, and I see what you were meaning there.

I suppose this is the downside of a framework since with frameworks and codebases there is often friction at the interface. It's unfortunate, in my view, because as it is, Phoenix generated code requires a lot of shaping up to pass relatively unopinionated Credo checks. I personally feel Phoenix has a high cost in this aspect.

But anyway, to help clarify, is Credo seen more as a side thing by the Elixir community? I will keep using it regardless, but I had the feeling it was encouraged by the "core community" and considered part of the core toolbelt. So perhaps I have some misunderstanding there in terms of its adoption and acceptance.

I do wish typespecs and documentation weren't seen as noise. I am still somewhat new to Elixir, but I've seen enough Elixir and Erlang code without typespecs (basically 99% of the code I've seen across two companies and additionally several libraries I've seen have no typespecs) and without documentation and the several problems that occur because of that to accept the noise of typespecs and documentation.

So is (3) at all encouraged? Or should I just not bother? (I do wish `@impl <behaviour_name>` was enforced by the compiler and `@impl true` never allowed. The latter only saves five characters in the most common case of implementing `GenServer`s.)

Lastly, I was also trying to clarify @chrismccord's comment because it wasn't clear, especially in context of your comment addressing a possible PR for (3).

**Comment by josevalim** [comment]
No one is saying documentation is noise. In fact, we do generate docs for context functions because we can do a good job at those. But generated documentation with placeholders or stubs will most likely become noise. Other than that, if you want Credo to enforce documentation, then you should go ahead and write it, and not have Phoenix generate something that is not useful. 

Similar goes for specs, if your team is not committed to maintaining them, then that's noise and outdated specs is worse than having no specs at all. The way to maintain them is with Dialyzer and I am skeptical that Dialyzer provides a good user experience for newcomers.

In other words, this is not about Credo at all. You are rather looking at those from the perspective of someone who wants to fully maintain those, and that's great, but you have to understand not everyone would make the same choices.

So no changes planned. :)

**Comment by bmitc** [comment]
So is a PR for (3) welcomed: yes or no?

Sorry to directly ask, again, but it's been left ambiguous. It was definitely welcomed and then no changes and not moving this direction was mentioned.

**Comment by josevalim** [comment]
Not for now, thanks!
